### PR TITLE
Add pronic number implementation

### PR DESCRIPTION
**Pull request by itsAkshayDubey**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Maths/IsPronic.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * therefore, the function will return true
+ *
+ * If number is not a product of two consecutive numbers, then it is not pronic
+ * therefore, the function will return false
+ * @param {number} number
+ * @returns {boolean}
+ */
+
+function isPronic(number) {
+
```
Comment (raklaptudirm) [inline Maths/IsPronic.js:22]: Add a check for whether the number is odd because a pronic number will never be odd.

File: Maths/IsPronic.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	for (var i = 0; i <= parseInt(Math.sqrt(number)); i++)
```
Comment (raklaptudirm) [inline Maths/IsPronic.js:23]: What is the use of the `parseInt` function here? It is used to convert a string into a number.
```js
parseInt("123") // returns 123
```

File: Maths/test/IsPronic.test.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { isPronic } from '../IsPronic'
+
+describe('Testing isPronic function', () => {
```
Comment (raklaptudirm) [inline Maths/test/IsPronic.test.js:3]: Add more test cases.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]
Format your code using `standard.js`. Read the `CONTRIBUTING.md` before contributing.

File: Maths/IsPronic.js (hunk 1 above)
Comment (itsAkshayDubey) [reply Maths/IsPronic.js:25]: Added check of odd number.

File: Maths/IsPronic.js (hunk 2 above)
Comment (itsAkshayDubey) [reply Maths/IsPronic.js]: Removed parseInt()

File: Maths/IsPronic.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * @param {number} number
+ * @returns {boolean}
+ */
+
+function isPronic (number) {
+  if (number % 2 === 1) {
+    return false
+  }
+
+  for (let i = 0; i <= Math.sqrt(number); i++) {
```
Comment (raklaptudirm) [inline Maths/IsPronic.js:26]: Instead of calculating the square root everytime, store it in a variable:
```js
const sqrt = Math.sqrt(number)
```

**Review by raklaptudirm — CHANGES_REQUESTED** [review]

File: Maths/IsPronic.js (hunk 4 above)
Comment (itsAkshayDubey) [reply Maths/IsPronic.js]: Added suggested changes to the code.
Thanks.

File: Maths/test/IsPronic.test.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  it('should return if the number is pronic or not for even number', () => {
+    const isPronicNumber = isPronic(2)
+    expect(isPronicNumber).toBe(true)
+  })
+
+  it('should return if the number is pronic or not for even number', () => {
+    const isPronicNumber = isPronic(4)
```
Comment (appgurueu) [inline Maths/test/IsPronic.test.js:10]: Move this to the first `it`.

File: Maths/IsPronic.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+function isPronic (number) {
+  const sqrt = Math.sqrt(number)
+
+  if (number % 2 === 1) {
+    return false
+  }
+
+  for (let i = 0; i <= sqrt; i++) {
```
Comment (appgurueu) [inline Maths/IsPronic.js:28]: This runs in O(sqrt n) time; it seems unnecessarily inefficient to me. Would checking around the square root not work? `Math.floor(sqrt) * Math.ceil(sqrt) == number` should hold for all pronic numbers.

Comment (itsAkshayDubey) [reply Maths/IsPronic.js]: Have a look.

**Comment by appgurueu** [comment]
The current implementation is very odd: (1) you don't need to check whether numbers are odd if your equation works (2) your equation is unnecessarily convoluted.

**Comment by itsAkshayDubey** [comment]
> The current implementation is very odd: (1) you don't need to check whether numbers are odd if your equation works (2) your equation is unnecessarily convoluted.

Refactored it, have a look.

**Comment by appgurueu** [comment]
Why use `Math.round` if you can use `Math.floor`/`Math.ceil`? I also failed to consider edge cases properly: `0` is pronic but `1` is not. The following should work:

```javascript
number !== 1 && (number % Math.ceil(Math.sqrt(number)) == 0)
```

**Comment by itsAkshayDubey** [comment]
> Why use `Math.round` if you can use `Math.floor`/`Math.ceil`? I also failed to consider edge cases properly: `0` is pronic but `1` is not. The following should work:
> 
> ```js
> if (number === 0) return true
> const sqrt = Math.sqrt(number)
> return sqrt % 1 !== 0 && Math.ceil(sqrt) * Math.floor(sqrt) === number
> ```
> 
> Please extend the tests to check all example numbers listed on https://oeis.org/A002378 - it's easy to mess up some edge cases. Also check that all other numbers in the range aren't false positives.

Added, have a look.

**Review by appgurueu — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]
